### PR TITLE
Sds: Ssl socket factory owns ContextConfig

### PR DESCRIPTION
**Pull request by qiwzhang**
*Description*:

This change is one of PR to support dynamic SDS.  The overall implementation PR is here <https://github.com/envoyproxy/envoy/pull/3700>

Changes in this PR
* use unique_ptr for ContextConfig object.  If the config is updated (secret is updated),  the same object will be used.
* ssl socket factory owns ContextConfig object so later it can be used to re-create Context if config is updated
* ssl socket factory adds un-used function onConfigUpdate() to illustrate how Context is created from updated config.  This function will be called in a later PR.

*Risk Level*:  Low

*Testing*:   Unit-test

### CODE REVIEW

**Comment by qiwzhang** [comment]
@lizan @PiotrSikora could you help to review it?

File: source/common/ssl/ssl_socket.h [hunk 1]
```diff
@@ -69,25 +69,38 @@ class SslSocket : public Network::TransportSocket,
...
-  ClientSslSocketFactory(const ClientContextConfig& config, Ssl::ContextManager& manager,
+  ClientSslSocketFactory(ClientContextConfigPtr config, Ssl::ContextManager& manager,
                          Stats::Scope& stats_scope);
 
   Network::TransportSocketPtr createTransportSocket() const override;
   bool implementsSecureTransport() const override;
 
 private:
+  // Will be called when config_ is changed
+  void onConfigUpdate();
```
Comment (lizan) [inline source/common/ssl/ssl_socket.h]: How this will be called? and tests? If the caller is not in this PR, include this in that PR.

Comment (qiwzhang) [reply source/common/ssl/ssl_socket.h]: Done.  onConfigUpdate is removed.

File: source/common/ssl/ssl_socket.h [hunk 2]
```diff
@@ -69,25 +69,32 @@ class SslSocket : public Network::TransportSocket,
...
+  ClientSslSocketFactory(ClientContextConfigPtr config, Ssl::ContextManager& manager,
                          Stats::Scope& stats_scope);
 
   Network::TransportSocketPtr createTransportSocket() const override;
   bool implementsSecureTransport() const override;
 
 private:
+  Ssl::ContextManager& manager_;
+  Stats::Scope& stats_scope_;
+  ClientContextConfigPtr config_;
```
Comment (PiotrSikora) [inline source/common/ssl/ssl_socket.h]: Why do we need to retain those? IT doesn't seem that those are used anywhere?

Comment (qiwzhang) [reply source/common/ssl/ssl_socket.h]: It will be used in a later PR to create a new context when context_config is updated with new secret.

Comment (qiwzhang) [reply source/common/ssl/ssl_socket.h]: Here is the code to use it
```
void ServerSslSocketFactory::onConfigUpdate() {
  ssl_ctx_ = manager_.createSslServerContext(stats_scope_, *config_, server_names_);
}

```

Comment (PiotrSikora) [reply source/common/ssl/ssl_socket.h]: OK, thanks!

**Pull request merged**